# Assignment 5 - Logistic Regression
(20 points)

### Add your name(s) and EIDs below
- Student Name: Siem Yonas
- Student UT EID: sby237
- Partner Name:
- Partner UT EID:

### Grading clarifications
1. Before you submit:
    - **Please do not change or remove any variable names** we give you for your answer to each question. You can use other variables if you wish, but the final answer must be stored in that variable.
    - **Please don't use any other libraries** besides those provided in `requirements.txt`, and those [built in to Python](https://docs.python.org/3.8/library/index.html).
2. We will also be double-checking the autograder, so you won't lose points if your formatting causes it to deduct points (but please try to follow the points under 1).
3. After submitting, you may see "All test cases passed!" for Public Tests. Please note that this doesn't mean you have correctly answered every question, as there are no public tests. We will be using a series of hidden tests to verify your answers are correct.

In [1]:
# Standard Headers
# You are welcome to add additional headers here if you wish, as long as they are
# built-in or included in requirements.txt
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Enable inline mode for matplotlib so that Jupyter displays graphs
%matplotlib inline

## Heart Dataset 

In this Assignment we will work with a dataset of patients. 
We have access to 303 patients' data. The features are listed below. 

In [2]:
# Your code here
heart_df = pd.read_csv("Heart.csv")
heart_df

Age  Sex     ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  \
0     63    1       typical     145   233    1        2    150      0   
1     67    1  asymptomatic     160   286    0        2    108      1   
2     67    1  asymptomatic     120   229    0        2    129      1   
3     37    1    nonanginal     130   250    0        0    187      0   
4     41    0    nontypical     130   204    0        2    172      0   
..   ...  ...           ...     ...   ...  ...      ...    ...    ...   
298   45    1       typical     110   264    0        0    132      0   
299   68    1  asymptomatic     144   193    1        0    141      0   
300   57    1  asymptomatic     130   131    0        0    115      1   
301   57    0    nontypical     130   236    0        2    174      0   
302   38    1    nonanginal     138   175    0        0    173      0   

     Oldpeak  Slope   Ca        Thal Target  
0        2.3      3  0.0       fixed     No  
1        1.5      2  3.0      normal    Yes  
2        2.6      2  2.0  reversable    Yes  
3        3.5      3  0.0      normal     No  
4        1.4      1  0.0      normal     No  
..       ...    ...  ...         ...    ...  
298      1.2      2  0.0  reversable    Yes  
299      3.4      2  2.0  reversable    Yes  
300      1.2      2  1.0  reversable    Yes  
301      0.0      2  1.0      normal    Yes  
302      0.0      1  NaN      normal     No  

[303 rows x 14 columns]

**Age:** The person’s age in years

**Sex:** The person’s sex (1 = male, 0 = female)

**ChestPain:** chest pain type

* Value 0: asymptomatic
* Value 1: atypical angina
* Value 2: non-anginal pain
* Value 3: typical angina

**RestBP:** The person’s resting blood pressure (mm Hg on admission to the hospital)

**Chol:** The person’s cholesterol measurement in mg/dl

**Fbs:** The person’s fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)
restecg: resting electrocardiographic results

* Value 0: showing probable or definite left ventricular hypertrophy by Estes’ criteria
* Value 1: normal
* Value 2: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)

**RestECG:** The person’s maximum heart rate achieved

**MaxHR:** Exercise induced angina (1 = yes; 0 = no)

**Oldpeak:** ST depression induced by exercise relative to rest (‘ST’ relates to positions on the ECG plot. See more here)

**Slope:** the slope of the peak exercise ST segment — 0: downsloping; 1: flat; 2: upsloping

* 0: downsloping; 
* 1: flat; 
* 2: upsloping

**Ca:** The number of major vessels (0–3)

**Thal:** A blood disorder called thalassemia Value 0: NULL (dropped from the dataset previously

* Value 1: fixed defect (no blood flow in some part of the heart)
* Value 2: normal blood flow
* Value 3: reversible defect (a blood flow is observed but it is not normal)

**Target:** Heart disease (1 = no, 0= yes)

# Q1
We want to use logistic regerssion to predict if a patient will have heart problems or not. The column "Target" in our datasets includes data about heart disease. If the patient had heart disease, the patient's "Target" value equals 1. Otherwise, "Target" equals 0.

Prepare your data set for predicting heart disease ("Target" column) by using 3 features:

* Age of the patient (Column **"Age"**)
* Gender of the patient (male or female - Column **"Sex"**)
* Cholestrol level of the patient (Column **"Chol"**) 

Split your data into 80% traning data and 20% test data. (**5 points**)

<!-- BEGIN QUESTION -->



In [3]:
# Put the features into an "X" array, and target variable into a "y" array.
X = heart_df[["Age", "Sex", "Chol"]]
y = heart_df["Target"]

In [4]:
# Split your "X" and "y" arrays into training and testing sets.
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=123)

print(X_train)

     Age  Sex  Chol
102   57    0   303
261   58    0   319
228   54    1   206
288   56    1   221
78    48    1   245
..   ...  ...   ...
106   59    1   177
83    68    1   274
17    54    1   239
230   52    0   196
98    52    1   201

[242 rows x 3 columns]


<!-- END QUESTION -->

# Q2

Generate a logistic regression model using your training data. 

Print out the accuracy if your model. (**10 points**)

<!-- BEGIN QUESTION -->



In [5]:
# Create a logistic regression model.
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()

# Fit the model to your training data.
model.fit(X_train, y_train)

LogisticRegression()

In [6]:
# Compute the accuracy of your model on the testing set.
from sklearn.metrics import accuracy_score

y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
accuracy

0.6885245901639344

<!-- END QUESTION -->

# Q3
Generate the classification report for your Logistic regresion model and interpret your results regarding precision, recall and f1-score. (**5 points - 2 for computation, 3 for interpretations**)


**Description**: In our test data, 73% of people classified as without heart disease did not have heart disease (highly accurate in classifying negatives), and 65% of people classified as with heart disease did have heart disease (good accuracy in classifying positives). 

67% of people without heart disease were correctly identified as without heart disease (A majority of all negative results are captured), and 71% of people with heart disease were correctly identified as with heart disease (A majority of all positive results are captured).

The F1-score for No and Yes are both high (0.70 and 0.68), incidating a good quality classifier.

<!-- BEGIN QUESTION -->



In [7]:
from sklearn.metrics import classification_report

report = classification_report(y_test,y_pred)
print(report)

              precision    recall  f1-score   support

          No       0.73      0.67      0.70        33
         Yes       0.65      0.71      0.68        28

    accuracy                           0.69        61
   macro avg       0.69      0.69      0.69        61
weighted avg       0.69      0.69      0.69        61



<!-- END QUESTION -->

